### ISSUE TITLE
DNS datagram buffer insufficient sizing

### ISSUE THREAD

**Issue by dmk23** [netty/netty#3971]
In running big DNS resolution loads with logging turned on I'm seeing lots of errors that seem to imply the datagram buffer gets cut off at 512 bytes. Below are a few examples.

Is this something that could be fixed by configuring within 4.1-beta5 or has to wait for new release?

```
2015-07-02 13:39:54,802 WARN  [epollEventLoopGroup-2-1] dns.DnsNameResolver: Unexpected exception:
io.netty.handler.codec.DecoderException: java.lang.IndexOutOfBoundsException: readerIndex(512) + length(4) exceeds writerIndex(512): PooledUnsafeDirectByteBuf(ridx: 512, widx: 512, cap: 4096)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:99)
        at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelReadNow(ChannelHandlerInvokerUtil.java:83)
        at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelRead(DefaultChannelHandlerInvoker.java:153)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:157)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:946)
        at io.netty.channel.epoll.EpollDatagramChannel$EpollDatagramChannelUnsafe.epollInReady(EpollDatagramChannel.java:581)
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:322)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:254)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:703)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
        at java.lang.Thread.run(Thread.java:744)
Caused by: java.lang.IndexOutOfBoundsException: readerIndex(512) + length(4) exceeds writerIndex(512): PooledUnsafeDirectByteBuf(ridx: 512, widx: 512, cap: 4096)
        at io.netty.buffer.AbstractByteBuf.checkReadableBytes(AbstractByteBuf.java:1139)
        at io.netty.buffer.AbstractByteBuf.readInt(AbstractByteBuf.java:592)
        at io.netty.buffer.AbstractByteBuf.readUnsignedInt(AbstractByteBuf.java:600)
        at io.netty.handler.codec.dns.DefaultDnsRecordDecoder.decodeRecord(DefaultDnsRecordDecoder.java:48)
        at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decodeRecords(DatagramDnsResponseDecoder.java:108)
        at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decode(DatagramDnsResponseDecoder.java:68)
        at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decode(DatagramDnsResponseDecoder.java:33)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:89)
        ... 10 more

2015-07-02 13:39:55,440 WARN  [epollEventLoopGroup-2-1] dns.DnsNameResolver: Unexpected exception:
io.netty.handler.codec.DecoderException: java.lang.IndexOutOfBoundsException: readerIndex: 524 (expected: 0 <= readerIndex <= writerIndex(512))
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:99)
        at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelReadNow(ChannelHandlerInvokerUtil.java:83)
        at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelRead(DefaultChannelHandlerInvoker.java:153)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:157)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:946)
        at io.netty.channel.epoll.EpollDatagramChannel$EpollDatagramChannelUnsafe.epollInReady(EpollDatagramChannel.java:581)
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:322)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:254)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:703)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
        at java.lang.Thread.run(Thread.java:744)
Caused by: java.lang.IndexOutOfBoundsException: readerIndex: 524 (expected: 0 <= readerIndex <= writerIndex(512))
        at io.netty.buffer.AbstractByteBuf.readerIndex(AbstractByteBuf.java:74)
        at io.netty.handler.codec.dns.DefaultDnsRecordDecoder.decodeRecord(DefaultDnsRecordDecoder.java:54)
        at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decodeRecords(DatagramDnsResponseDecoder.java:108)
        at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decode(DatagramDnsResponseDecoder.java:68)
        at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decode(DatagramDnsResponseDecoder.java:33)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:89)
        ... 10 more

2015-07-02 13:39:55,720 WARN  [epollEventLoopGroup-2-1] dns.DnsNameResolver: Unexpected exception:
io.netty.handler.codec.DecoderException: java.lang.IndexOutOfBoundsException: readerIndex: 518 (expected: 0 <= readerIndex <= writerIndex(512))
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:99)
        at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelReadNow(ChannelHandlerInvokerUtil.java:83)
        at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelRead(DefaultChannelHandlerInvoker.java:153)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:157)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:946)
        at io.netty.channel.epoll.EpollDatagramChannel$EpollDatagramChannelUnsafe.epollInReady(EpollDatagramChannel.java:581)
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:322)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:254)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:703)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
        at java.lang.Thread.run(Thread.java:744)
Caused by: java.lang.IndexOutOfBoundsException: readerIndex: 518 (expected: 0 <= readerIndex <= writerIndex(512))
        at io.netty.buffer.AbstractByteBuf.readerIndex(AbstractByteBuf.java:74)
        at io.netty.handler.codec.dns.DefaultDnsRecordDecoder.decodeRecord(DefaultDnsRecordDecoder.java:54)
        at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decodeRecords(DatagramDnsResponseDecoder.java:108)
        at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decode(DatagramDnsResponseDecoder.java:68)
        at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decode(DatagramDnsResponseDecoder.java:33)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:89)
        ... 10 more
```

**Comment by trustin** [comment]
That's interesting. `DnsNameResolver` by default uses a 4096-byte buffer to receive a datagram packet.  Are you sure you did not call `DnsNameResolver.setMaxPayloadSize(512)` or the server actually sent a truncated response?

**Comment by trustin** [comment]
If the server sent a truncated response, `DnsNameResolver` should not attempt to decode it but just fail with a better explanation, though.

**Comment by dmk23** [comment]
I did not set the payload size. Not sure if the servers are sending the truncated responses or they get truncated somewhere along the way in Netty. Here is my resolver config:

```
resolver = new DnsNameResolver(group.next(), EpollDatagramChannel.class, PUBLIC_DNS)
  .setTtl(0, 0) 
  .setQueryTimeoutMillis(3000) 
  .setMaxTriesPerQuery(3)
  .setNegativeTtl(3600);
```

This issue is probably highly interrelated with the other ones I've just reported. As far as error handling there are many open issues to consider: https://github.com/netty/netty/issues/created_by/dmk23
